fair_drop/suspicious.py

```diff
@@ -0,0 +1,180 @@
+    respect_retry_after_header=True,
+)
+adapter = HTTPAdapter(max_retries=retry_strategy)
+scraper.mount("https://", adapter)
+scraper.mount("http://", adapter)
+
+
+def is_nft_suspicious(nft_url):
+    logging.debug(f"Scraping NFT with link: {nft_url}")
+    res = scraper.get(nft_url)
+
+    if res.status_code == 404:  # NFT not found
+        logging.info("NFT not found. Probably reached the end of a collection")
+        return None, None
+    if res.status_code == 200:
+        soup = BeautifulSoup(res.text, "html.parser")
+        collection_name = soup.find(class_="item--collection-detail").text
+        owner = soup.find(class_="AccountLink--ellipsis-overflow").text.replace(
+            "Owned by\xa0", ""
+        )
+        data = {
+            "collection": args.collection_address,
+            "collection_name": collection_name,
+            "blockchain": "ethereum",
+            "url": nft_url,
+            "owner": owner,
+        }
+        if res.text.find("Reported for suspicious") > 0:
+            logging.info(f"Found suspicious NFT of URL {nft_url}")
+            return True, data
+        else:
+            return False, data
+
+
+OPENSEA_BASE_URL = (
+    "https://opensea.io/assets/ethereum/"  # TODO adjust for other blockchains
+)
+
+
+def list_collection_nfts_urls(collection_address):
+    """List all OpenSea URLs of NFTs in a collection
+
+    Args:
+        collection_address (string): NFT collection address
+
+    Returns:
+        array: list of the OpenSea URLs of NFTs
+    """
+    # ! This is just a mock function. It is to be replaced with a call to the OpenSea API
+    nft_urls = []
+    for i in range(0, 9999):
+        nft_urls.append(f"{OPENSEA_BASE_URL}{collection_address}/{i}")
+    return nft_urls
+
+
+def scrape_all_collection_suspicious_nfts(collection_address):
+    collection_nfts_urls = list_collection_nfts_urls(collection_address)
+    logging.info(f"Collection contains {len(collection_nfts_urls)} NFTs")
+    collection_cache = load_scrape_cache(collection_nfts_urls)
+    logging.info(f"Found {len(collection_cache)} NFTs in collection cache")
+    # TODO removed scraped URLs from the list to be scraped
+    for index, nft in collection_cache.iterrows():
+        if (
+            nft["url"] in collection_nfts_urls
+        ):  # TODO add an expiry rule, depending on date of last scraped
+            logging.debug(f"NFT to be scraped already in cache. Skipping {nft['url']}")
+            collection_nfts_urls.remove(nft["url"])
+    logging.info(f"Scraping a list of {len(collection_nfts_urls)} NFTs")
+    nft_urls_batches = [
+        collection_nfts_urls[i : i + args.batch_size]
+        for i in range(0, len(collection_nfts_urls), args.batch_size)
+    ]
+    for index, batch in enumerate(nft_urls_batches):
+        print(f"Scraped {index * args.batch_size} NFT URLs so far")
+        with Pool(5) as p:
+            # ! Multiple return values
+            results = p.map(is_nft_suspicious, batch)
```

Comment:
`$ py3 fair_drop/suspicious.py -c 0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d`
This suddenly ends in an error:

```
INFO:root:Collection contains 9999 NFTs
INFO:root:Found 0 NFTs in collection cache
INFO:root:Scraping a list of 9999 NFTs
Scraped 0 NFT URLs so far
Scraped 50 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/71
Scraped 100 NFT URLs so far
Scraped 150 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/181
Scraped 200 NFT URLs so far
Scraped 250 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/290
Scraped 300 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/313
Scraped 350 NFT URLs so far
Scraped 400 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/403
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/404
Scraped 450 NFT URLs so far
Scraped 500 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/511
Scraped 550 NFT URLs so far
Scraped 600 NFT URLs so far
Scraped 650 NFT URLs so far
Scraped 700 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/731
Scraped 750 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/753
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/775
Scraped 800 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/842
Scraped 850 NFT URLs so far
Scraped 900 NFT URLs so far
Scraped 950 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/993
Scraped 1000 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/1013
Scraped 1050 NFT URLs so far
Scraped 1100 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/1100
Scraped 1150 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/1181
Scraped 1200 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/1207
Scraped 1250 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/1277
Scraped 1300 NFT URLs so far
Scraped 1350 NFT URLs so far
INFO:root:Found suspicious NFT of URL https://opensea.io/assets/ethereum/0xbc4ca0eda7647a8ab7c2061c2e118a18a936f13d/1364
Scraped 1400 NFT URLs so far
Scraped 1450 NFT URLs so far
Traceback (most recent call last):
  File "/home/xxx/honestnft-shenanigans/fair_drop/suspicious.py", line 180, in <module>
    scrape_all_collection_suspicious_nfts(args.collection_address)
  File "/home/xxx/honestnft-shenanigans/fair_drop/suspicious.py", line 143, in scrape_all_collection_suspicious_nfts
    results = p.map(is_nft_suspicious, batch)
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 771, in get
    raise self._value
multiprocessing.pool.MaybeEncodingError: Error sending result: '<multiprocessing.pool.ExceptionWithTraceback object at 0x7fddcef4ba30>'. Reason: 'TypeError("cannot pickle 'zlib.Decompress' object")'

```

Reply:
hmm :thinking: this is new to me... I'll see if I can reproduce it

Reply:
@Barabazs I did experience this error on my side too, but the problem is that I can't reproduce it reliably.

I imagine that my `is_nft_suspicious` function fails. Maybe due to rate limiting, or is left hanging without a response due to network problem :thinking:

Reply:
I just ran into it again.